src/export/FSHExporter.ts

```diff
@@ -63,150 +75,190 @@ export class FSHExporter {
-    const files: Map<string, string[]> = new Map();
+  private groupByType(): Map<string, Exportable[]> {
+    const files: Map<string, Exportable[]> = new Map();
 
     // Group profiles and examples of those profiles into individual files
     this.fshPackage.profiles.forEach(profile => {
-      files.set(`${profile.name}.fsh`, [profile.toFSH()]);
+      files.set(`${profile.name}.fsh`, [profile]);
     });
-    this.fshPackage.instances.forEach(instance => {
+
+    files.set('instances.fsh', []);
+
+    const [inlineInstances, nonInlineInstances] = partition(
+      this.fshPackage.instances,
+      i => i.usage === 'Inline'
+    );
+    // If a non-inline instance is an example of a profile, it is written to the file
+    // for that profile. Otherwise it is written to instances.fsh
+    nonInlineInstances.forEach(instance => {
       if (instance.usage === 'Example' && files.has(`${instance.instanceOf}.fsh`)) {
-        files.get(`${instance.instanceOf}.fsh`).push(instance.toFSH());
-      } else if (files.has('instances.fsh')) {
-        files.get('instances.fsh').push(instance.toFSH());
+        files.get(`${instance.instanceOf}.fsh`).push(instance);
       } else {
-        files.set('instances.fsh', [instance.toFSH()]);
+        files.get('instances.fsh').push(instance);
       }
     });
+    // Inline instances are written to the file they are used in, if they are only used
+    // in one spot. Otherwise they go to instances.fsh
+    inlineInstances.forEach(instance => {
+      const usedIn = this.inlineInstanceUsedIn(instance, files);
+      usedIn.length === 1
+        ? files.get(usedIn[0]).push(instance)
+        : files.get('instances.fsh').push(instance);
+    });
+
+    // Invariants are written to the same file as the profile they are used in, if they
+    // are written in one spot. Otherwise they go to invariants.fsh.
+    files.set('invariants.fsh', []);
+    this.fshPackage.invariants.forEach(invariant => {
+      const usedIn = this.invariantUsedIn(invariant, files);
+      usedIn.length === 1
+        ? files.get(usedIn[0]).push(invariant)
+        : files.get('invariants.fsh').push(invariant);
+    });
 
     // All other artifacts are grouped by category
-    files.set('aliases.fsh', [this.fshPackage.aliases.map(a => a.toFSH()).join(EOL)]);
-    files.set(
-      'extensions.fsh',
-      this.fshPackage.extensions.map(extension => extension.toFSH())
-    );
-    files.set('terminology.fsh', [
-      ...this.fshPackage.valueSets.map(valueSet => valueSet.toFSH()),
-      ...this.fshPackage.codeSystems.map(codeSystem => codeSystem.toFSH())
-    ]);
-
-    files.set(
-      'invariants.fsh',
-      this.fshPackage.invariants.map(invariant => invariant.toFSH())
-    );
-    files.set(
-      'mappings.fsh',
-      this.fshPackage.mappings.map(mapping => mapping.toFSH())
-    );
+    files.set('aliases.fsh', this.fshPackage.aliases);
+    files.set('extensions.fsh', this.fshPackage.extensions);
```

Comment:
I can't remember if we already discussed this, but would it make sense to do a similar sorting for extensions, where they are put in the file of a profile where they're used if they're only used on one profile? Or does it make more sense to keep them separate since they are more of their own concept, independent from profiles?

Reply:
I think it makes sense to leave extensions to themselves, since extensions are meant to stand more alone than something like an inline instance or an invariant. They get their own page in an IG.

Reply:
I buy it!